### ISSUE TITLE
SLS signature verification failure

### ISSUE THREAD

**Issue by williamhector** [onelogin/php-saml#243]
This may or may be related to Issue #225 
I am using version 2.11.0, but haven't kept an eye on which versions I've used in the past.

In `OneLogin_Saml2_Utils->validateBinarySign`, if `$retrieveParametersFromServer` is FALSE and the IdP's logout request contains lowercase url-encoded values, eg `%2f` instead of `%2F`, as Azure does, then when it calls `urlencode()` to rebuild the `$signedQuery`, it results in a query string that has uppercase url-encoded values, obviously causing the signature verification to fail.

I can set `$retrieveParametersFromServer` to true, which resolves the issue by getting the `$signedQuery` from `$_SERVER['QUERY_STRING']` instead of `$_GET`, however I'm not certain if `$retrieveParametersFromServer` is actually meant for a different purpose. 

Is there another purpose for `$retrieveParametersFromServer`, which could potentially cause other issues, or is that what it was built for?
I'd assume there is another reason, or else why isn't it the default option, since it seems less error prone...

**Comment by pitbulk** [comment]
In order to fix the ADFS issue you have an specific attribute: [lowercaseUrlencoding](https://github.com/onelogin/php-saml/blob/master/advanced_settings_example.php#L101).

The$retrieveParametersFromServer parameter [was introduced in 2014](https://github.com/onelogin/php-saml/commit/6a828bf15502e9080ec4137964c0b310148bebf2), also to solve ADFS issues, but right now can't remember what exactly it was.

**Comment by williamhector** [comment]
Thanks for the tip. However, from what I understand so far it might be better to always do signature validation against the unaltered $_SERVER['QUERY_STRING'] instead of re-encoding from the $_GET vars.

I'd rather not use that lowercaseUrlencoding setting because our IdPs could change theirs without notice.

Because of this and some other things I've noticed, I get the sense that this project might need a major, breaking version change at some point. For example, I noticed that there is both an "x509cert" and "x509certMulti" IdP setting, but the code base could be simplified if it only supported the array setting.
I'd love to be able to help if that is a goal.
